network/src/sync/master.rs

```diff
@@ -298,8 +311,13 @@ impl<S: Storage + Send + Sync + 'static> SyncMaster<S> {
         let early_blocks = Self::order_block_hashes(&blocks[..]);
         let early_blocks_count = early_blocks.len();
 
-        let ledger = &self.node.expect_sync().consensus.ledger;
-        let block_order: Vec<BlockHeaderHash> = early_blocks.into_iter().filter(|x| !ledger.is_canon(x)).collect();
+        let early_block_states = self.node.storage.get_block_states(&early_blocks[..]).await?;
+        let block_order: Vec<_> = early_blocks
+            .iter()
+            .zip(early_block_states.iter())
+            .filter(|(_, status)| matches!(status, snarkos_storage::BlockStatus::Unknown))
+            .map(|(hash, _)| hash.clone())
```

Comment:
No need for the `clone` here if you do `into_iter` twice, which should be possible here.

Reply:
i'll do that, but `Digest` cloning is usually as cheap as move-by-value, which we do anyways. `Digest`s in the wild are < 64 bytes.